li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaInstrumentedProducerImpl.java

```diff
@@ -444,15 +446,38 @@ private boolean proceedClosing() {
     if (isClosed()) {
       return false;
     }
-    delegateLock.writeLock().lock();
-    try {
+    synchronized (closeLock) {
       if (isClosed()) {
-        return false;
+        return false; //someone beat us to it
+      }
+      int holds = delegateLock.getReadHoldCount(); //this is for our thread
+      ReentrantReadWriteLock.ReadLock readLock = delegateLock.readLock();
+      ReentrantReadWriteLock.WriteLock writeLock = delegateLock.writeLock();
+      if (holds > 0) { //do we own a read lock ?
+        for (int i=0; i<holds; i++) {
+          readLock.unlock();
+        }
+        //at this point we no longer hold a read lock, but any number of other
+        //readers/writers may slip past us
+      }
+      try {
+        writeLock.lock(); //wait for a write lock
+        try {
+          if (isClosed()) {
+            return false; //some other writer may have beaten us again
+          }
+          closedAt = System.currentTimeMillis();
+          return true;
+        } finally {
+          writeLock.unlock();
+        }
+      } finally {
+        if (holds > 0) { //restore our read lock holds (if we had any)
+          for (int i=0; i<holds; i++) {
+            readLock.lock();
```

Comment:
Does grab write lock in recreateDelegate have any impact here?

Reply:
shouldnt, we wait for the write lock separately later

Reply:
recreateDelegate() should check for closed, now that i think about it

Reply:
the general answer - close() terminates this client. it _may_ race with recreateDelegate(), but the result should be "just" wasted effort in creating a new delegate and then immediately closing it.